modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,153 @@
+//            EitherT.fromEither(weightByStardust(distribution))
+//            EitherT.fromEither {
+//              Kleisli(weightBySoftStaking(epochProgress))
+//                .compose(weightByDTM)
+//                .compose(weightByStardust)
+//                .apply(distribution)
+//            }
+//          }
+          .flatMap { distribution =>
+            EitherT.fromEither {
+              distribution.toNel.traverse {
+                case (address, reward) =>
+                  PosLong
+                    .from(reward.value)
+                    .map(reward => RewardTransaction(address, TransactionAmount(reward)))
+                    .left
+                    .map[RewardsError](NumberRefinementPredicatedFailure)
+              }.map(_.toNes.toSortedSet)
+            }
+          }
+          .flatTap(txs => EitherT.liftF(logger.info(s"Minted reward transactions: $txs")))
+          .valueOrF { error =>
+            val logError = error match {
+              case LastEpochExceeded => logger.warn("Last epoch exceeded. Won't mint anymore")
+              case err @ NumberRefinementPredicatedFailure(_) =>
+                logger.error(err)("Unhandled refinement predicate failure")
+              case IgnoredAllAddressesDuringWeighting =>
+                logger.warn("Ignored all the addresses from distribution so weights won't be applied")
+            }
+
+            logError.as(SortedSet.empty[RewardTransaction])
+          }
+
+      def getRewardsPerSnapshot(epochProgress: EpochProgress): Either[RewardsError, Amount] =
+        config.rewardsPerEpoch.toNel.toList
+          .sliding(2)
+          .map { l =>
+            (l.head, l.tail.headOption.getOrElse(l.head))
+          }
+          .find {
+            case ((_, _), (curr, _)) => epochProgress.value <= curr.value
+          }
+          .map {
+            case ((prevOrdinal, prevR), (_, currR)) =>
+              if (epochProgress.value <= prevOrdinal.value) prevR
+              else currR
+          }
+          .toRight(LastEpochExceeded)
+
+      private[rewards] def createSharedDistribution(
+        epochProgress: EpochProgress,
+        snapshotReward: Amount,
+        addresses: NonEmptySet[Address]
+      ): EitherT[F, RewardsError, NonEmptyMap[Address, Amount]] = {
+
+        val (quotient, remainder) = snapshotReward.coerce.toLong /% addresses.length.toLong
+
+        val quotientDistribution = addresses.toNonEmptyList.map(_ -> NonNegLong.unsafeFrom(quotient)).toNem
+
+        EitherT.liftF {
+          Random.scalaUtilRandomSeedLong(epochProgress.coerce).flatMap { random =>
+            random
+              .shuffleList(addresses.toNonEmptyList.toList)
+              .map { shuffled =>
+                shuffled.take(remainder.toInt)
+              }
+              .map { top =>
+                top.toNel
+                  .fold(quotientDistribution) { topNel =>
+                    topNel.map(_ -> NonNegLong(1L)).toNem |+| quotientDistribution
```

Comment:
I think that combining rewards isn't correct. If there are different rewards from different programs to the same address I believe it should be separate transactions instead of a single transaction for a combined amount.